src/unix_process/unix_process_runtime.c

```diff
@@ -85,25 +85,33 @@ void halt(char *format, ...)
     exit(2);
 }
 
-#ifndef __aarch64__
-static heap allocate_tagged_region(heap h, u64 tag)
-{
-    u64 size = 256 * MB;
-    void *region = mmap(pointer_from_u64(tag << VA_TAG_OFFSET),
-                        size, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_ANON | MAP_PRIVATE, -1, 0);
-    return (heap)create_id_heap(h, h, u64_from_pointer(region), size, 1, false);
-}
-#else
 struct tagheap {
     struct heap h;
     heap mh;
-    u64 vtag;
+    u64 tag;
 };
 
+#ifndef __aarch64__
+
+static inline u64 tag_alloc(heap h, bytes length)
```

Comment:
I suggest that malloc_alloc and malloc_free are also made to allocate an additional (zeroed) byte like the tagged methods here - in essence making every alloc a tagged alloc. Existing code expects to be able to check the tag of any value which may include values that weren't allocated from a tagged region, like string buffers. This works by default with address tagging, but without a padded alloc here it could result in a dereference outside the allocated area, leading either to a fault or an incorrect tag being read.

Reply:
OK, I made malloc_alloc() and malloc_free() arch-specific (with the x864_64 variants allocating extra space for the tag), while tag_alloc() and tag_dealloc() are now arch-independent (the arch-specific part is in the tag() inline function).
The extra space for the tag has to be 8 bytes, not 1, otherwise mks segfaults somewhere within filesystem_write_linear(). While I haven't identified the root cause of this, I think it's wise to ensure in any case that the pointer returned by the alloc function is aligned to 64 bits (even though Intel CPUs support unaligned accesses, there may be a performance penalty).